### PR TITLE
Initial commit of auth_mechanism 

### PR DESCRIPTION
**Pull request by rhysmeister**
##### SUMMARY

Add auth_mechanism param to modules. For issue https://github.com/ansible-collections/community.mongodb/issues/177

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
mongodb_common.py

### CODE REVIEW

**Review by cognifloyd — APPROVED** [review]
I can't test this with an X509 cert, but from reading, it looks good.

**Comment by cognifloyd** [comment]
https://github.com/ansible-collections/community.mongodb/pull/179/files#diff-09e1d4bab336fc280bf4e7dc2bcb29e9L333-L342

Since MONGODB-X509 doesn't require login_password, we're going to need special handling in mongodb_user and possibly other modules. This assumes username/password.

**Comment by cognifloyd** [comment]
The requirement for both login_user and login_password also changes with auth_mechanism, because MONGODB-X509 allows username, but password is not needed.

**Comment by rhysmeister** [comment]
> https://github.com/ansible-collections/community.mongodb/pull/179/files#diff-09e1d4bab336fc280bf4e7dc2bcb29e9L333-L342
> 
> Since MONGODB-X509 doesn't require login_password, we're going to need special handling in mongodb_user and possibly other modules. This assumes username/password.

Should be fine for most modules as we only authenticate with a user if one is provided. Just need to watch out that the ~/.mongodb.cnf file doesn't exist.

File: plugins/modules/mongodb_user.py [hunk 1]
```diff
@@ -215,10 +215,13 @@ def user_add(module, client, db_name, user, password, roles):
         # Should allow us to create the first user. So we assume this is the case
         # Might be able to do something with db.getUsers() if this approach needs
         # a rethink
-        if excep.code == 13:  # Unauthorized
-            exists = False
+        if hasattr(excep, 'code'):
+            if excep.code == 13:  # Unauthorized
+                exists = False
+            else:
+                raise excep
```
Comment (cognifloyd) [inline plugins/modules/mongodb_user.py:222]:
```suggestion
        if hasattr(excep, 'code') and excep.code == 13:  # Unauthorized
            exists = False
```

File: plugins/modules/mongodb_user.py [hunk 2]
```diff
@@ -215,10 +215,13 @@ def user_add(module, client, db_name, user, password, roles):
...
         else:
-            raise Exception
+            raise excep
```
Comment (cognifloyd) [inline plugins/modules/mongodb_user.py:224]:
```suggestion
            raise
```

**Review by cognifloyd — CHANGES_REQUESTED** [review]
minor code style change

**Comment by cognifloyd** [comment]
> Should be fine for most modules as we only authenticate with a user if one is provided. Just need to watch out that the ~/.mongodb.cnf file doesn't exist.

Ah. Re-reading the doc you linked to:

> _Changed in version 3.4:_ When connected to MongoDB >= 3.4 the username is no longer required.

Since the username issue is only for Mongo >=2.6,<3.4, assuming that `MONGODB-X509` auth will not pass a `login_username` is fairly safe because those versions are so old.

**Comment by rhysmeister** [comment]
> > Should be fine for most modules as we only authenticate with a user if one is provided. Just need to watch out that the ~/.mongodb.cnf file doesn't exist.
> 
> Ah. Re-reading the doc you linked to:
> 
> > _Changed in version 3.4:_ When connected to MongoDB >= 3.4 the username is no longer required.
> 
> Since the username issue is only for Mongo >=2.6,<3.4, assuming that `MONGODB-X509` auth will not pass a `login_username` is fairly safe because those versions are so old.

It does pass a username if one is provided. Spotted this during my testing because I had a .mongodb.cnf file present.

File: plugins/modules/mongodb_user.py (hunk 2 above)
Comment (rhysmeister) [reply plugins/modules/mongodb_user.py:224]: Bad habit from my Java days. I'll probably be doing this everywhere.

**Comment by rhysmeister** [comment]
@cognifloyd @fischerscode 

I'll run the k8 test again before merging this if there's no more input from you guys?

**Comment by rhysmeister** [comment]
Retested. Looks good.

```
(venv) (base) Rhyss-MacBook-Pro:mongodb rhyscampbell$ ansible localhost -m mongodb_user -a "login_host=localhost login_port=27017 login_database='$external' database='admin' password='secret' ssl=true ssl_ca_certs=/Users/rhyscampbell/Documents/git/mongodb-kub/tests/ansible-operator/ca.crt ssl_certfile=/Users/rhyscampbell/Documents/git/mongodb-kub/tests/ansible-operator/tls.key auth_mechanism=MONGODB-X509 name="test" state=present connection_options='tlsAllowInvalidHostnames=true'"
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible
engine, or trying out features under development. This is a rapidly changing source of code and can become unstable at any point.
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: Module did not set no_log for update_password
localhost | CHANGED => {
    "changed": true,
    "user": "test"
}
(venv) (base) Rhyss-MacBook-Pro:mongodb rhyscampbell$ ansible localhost -m mongodb_user -a "login_host=localhost login_port=27017 login_database='$external' database='admin' password='secret' ssl=true ssl_ca_certs=/Users/rhyscampbell/Documents/git/mongodb-kub/tests/ansible-operator/ca.crt ssl_certfile=/Users/rhyscampbell/Documents/git/mongodb-kub/tests/ansible-operator/tls.key auth_mechanism=MONGODB-X509 name="test2" state=present connection_options='tlsAllowInvalidHostnames=true'"
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible
engine, or trying out features under development. This is a rapidly changing source of code and can become unstable at any point.
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: Module did not set no_log for update_password
localhost | CHANGED => {
    "changed": true,
    "user": "test2"
}
(venv) (base) Rhyss-MacBook-Pro:mongodb rhyscampbell$ ansible localhost -m mongodb_user -a "login_host=localhost login_port=27017 login_database='$external' database='admin' password='secret' ssl=true ssl_ca_certs=/Users/rhyscampbell/Documents/git/mongodb-kub/tests/ansible-operator/ca.crt ssl_certfile=/Users/rhyscampbell/Documents/git/mongodb-kub/tests/ansible-operator/tls.key auth_mechanism=MONGODB-X509 name="test3" state=present connection_options='tlsAllowInvalidHostnames=true'"
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible
engine, or trying out features under development. This is a rapidly changing source of code and can become unstable at any point.
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: Module did not set no_log for update_password
localhost | CHANGED => {
    "changed": true,
    "user": "test3"
}
```

**Review by cognifloyd — APPROVED** [review]
Is the k8s test running in CI? It looks like not yet.

In any case I think this looks good. If you're satisfied that this isn't a draft any more, merge it.

**Comment by rhysmeister** [comment]
> Is the k8s test running in CI? It looks like not yet.
> 
> In any case I think this looks good. If you're satisfied that this isn't a draft any more, merge it.

No, not yet.

**Pull request merged**